KEYCLOAK-11862 Add Sync mode option

Martin-Idel-SI opened the pull request:
This PR introduces the sync mode option.
To make an initial review of the functionality and implementation strategy simpler, I only adapted the oidc frontend and did not adapt the tests (nor write new tests).

Regarding the proposal in https://issues.redhat.com/browse/KEYCLOAK-11862?_sscc=t this PR introduces the "legacy", "import" and "force" mode (since the "owner" mode requires additions to the user model and therefore introduces further complexity, I would rather do it in a future PR). It also adapts all existing mappers to honour those flags.

The sync mode observes the following behaviour:
- Old SPI extensions will continue to work and don't need any code changes. Their behaviour in "legacy" mode will be exactly the same. The behaviour in "import" mode will be always correct while the behaviour in "force" mode will only be correct by default if the legacy implementation would always update
- In order to get correct "force" behaviour for custom SPIs where it is not automatically correct, the authors of those SPIs only have to override one method. 
- No change to the database model (we store the sync mode in the config map).
- When existing IDPs get imported, everything will be set to legacy and they should work exactly as before (no config change necessary)
- New IDPs will default to "import" (new mappers to "no override", keeping the same sync mode as the IDP)

Once the general strategy is deemed good to go, I will update the PR according to the contribution guidelines:
- [x] Update all IDP frontends
- [x] Update functional tests
- [x] Include new functional/integration tests
- [x] Includes documentation
- [x] Squash to one commit

Martin-Idel-SI committed:
KEYCLOAK-11862 Add Sync mode option

- Store in config map in database and model
- Expose the field in the OIDC-IDP
- Write logic for import, force and legacy mode
- Show how mappers can be updated keeping correct legacy mode
- Show how mappers that work correctly don't have to be modified
- Log an error if sync mode is not supported

Fix updateBrokeredUser method for all mappers

- Allow updating of username (UsernameTemplateMapper)
- Delete UserAttributeStatementMapper: mapper isn't even registered
  Was actually rejected but never cleaned up: https://github.com/keycloak/keycloak/pull/4513
  The mapper won't work as specified and it's not easy to tests here
- Fixup json mapper
- Fix ExternalKeycloakRoleToRoleMapper:
  Bug: delete cannot work - just delete it. Don't fix it in legacy mode

Rework mapper tests

- Fix old tests for Identity Broker:
  Old tests did not work at all:
  They tested that if you take a realm and assign the role,
  this role is then assigned to the user in that realm,
  which has nothing to do with identity brokering
  Simplify logic in OidcClaimToRoleMapperTests
- Add SyncMode tests to most mappers
  Added tests for UsernameTemplateMapper
  Added tests to all RoleMappers
  Add test for json attribute mapper (Github as example)
- Extract common test setup(s)
- Extend admin console tests for sync mode

Signed-off-by: Martin Idel <external.Martin.Idel@bosch.io>

Martin-Idel-SI commented:
I now added all tests I think we should have at least (including one admin console test).
Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/914 

I left the commits as is to make reviewing maybe a little bit easier, I'll squash once all issues are resolved.

Martin-Idel-SI commented:
I also found that the OIDCAdvancedClaimToRole mapper is still missing its SAML counterpart. Would it be okay to also add it here?

mposolda reviewed (changes requested):
@Martin-Idel-SI Thanks for this work and PR! It is great stuff. And also thanks a lot for the great test coverage you added!

I've added few inline comments for the model layer with some suggested changes.

mposolda commented:
Regarding missing OIDCAdvancedClaimToRole  for SAML, I suggest separate JIRA and PR. This PR is already quite big :)

stianst commented on server-spi-private/src/main/java/org/keycloak/broker/provider/IdentityProviderMapperSyncModeDelegate.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public static IdentityProviderSyncMode combineIdpAndMapperSyncMode(IdentityProviderSyncMode syncMode, IdentityProviderMapperSyncMode mapperSyncMode) {
+        switch (mapperSyncMode) {
```
This can be changed to a single line:
`return IdentityProviderSyncMode.INHERIT.equals(mapperSyncMode) ? syncMode : mapperSyncMode `

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:
```diff
@@ -1287,25 +1291,31 @@ public void addIdentityProvider(IdentityProviderModel identityProvider) {
+        em.flush();
+    }
+
+    private void fillEntity(IdentityProviderModel identityProvider, IdentityProviderEntity entity) {
```
Not super keen on this refactoring - as it makes it more costly to review this. @mposolda did you review this properly?

Martin-Idel-SI replied:
This was just a trivial extraction, but since I'm removing the changes in RealmAdapter, I'll also remove the refactoring.

mposolda replied:
@stianst I think that this PR doesn't need to change anything in the RealmAdapter class at all if some of my other comments are applied - this means using just field from "config" map instead of adding additional field into the IdentityProviderModel / IdentityProviderMapperModel classes.

stianst replied:
Sounds good

stianst commented on server-spi-private/src/main/java/org/keycloak/broker/provider/IdentityProviderMapper.java:
```diff
@@ -17,25 +17,31 @@
-import org.keycloak.models.KeycloakSession;
-import org.keycloak.models.RealmModel;
-import org.keycloak.models.UserModel;
+import org.keycloak.models.*;
```
Please don't use '*' imports

stianst commented on server-spi-private/src/main/java/org/keycloak/broker/provider/IdentityProviderMapper.java:
```diff
@@ -17,25 +17,31 @@
     String getDisplayCategory();
     String getDisplayType();
 
+    default Set<IdentityProviderSyncMode> supportedSyncModes() {
```
Perhaps it would be cleaner to do:
`boolean supportsSyncMode(IdentityProviderSyncMode syncMode)`

stianst commented on server-spi/src/main/java/org/keycloak/models/IdentityProviderModel.java:
```diff
@@ -64,6 +66,8 @@
 
     private String displayName;
 
+    private IdentityProviderSyncMode syncMode;
```
+1

Martin-Idel-SI replied:
I removed the field and relied on the config (the getter will then provide the default value "LEGACY" in case no value exists). With this change I removed all changes to the representation end entity and there converters.

Martin-Idel-SI commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountFormServiceTest.java:
```diff
@@ -121,20 +122,24 @@ public void configureTestRealm(RealmRepresentation testRealm) {
         testRealm.addIdentityProvider(IdentityProviderBuilder.create()
                                               .providerId("github")
                                               .alias("github")
+                                              .syncMode(IdentityProviderSyncMode.IMPORT)
```
I removed this line where possible. In fact, this means that now my test changes are limited to the following places:
- classes within `org.keycloak.testsuite.broker` package: Most of the tests that don't need the sync mode are left as is, but the broker configurations all need to be adapted due to the nature of the inheritance structure. 
- classes outside `org.keycloak.testsuite.broker` are only adapted where I added tests (IdentityProviderTest, admin console tests and CLI-Test)

Is this enough in your view?

mposolda replied:
Yes, Thanks!

Martin-Idel-SI commented:
I tried to implement your review suggestions. Please let me know if the changes are sufficient @mposolda

mposolda commented:
@Martin-Idel-SI Thanks a lot for the updates! I suggest to squash the commits and rebase to latest master. That will hopefully help with the last failed test (MigrationModelTest).

mposolda commented:
@Martin-Idel-SI The failing MigrationModelTest is unrelated to your PR. I am looking at that one. Will let you know.

Martin-Idel-SI commented:
@mposolda Okay, so I rebased onto master and squashed to one commit, let's see whether that fixes the test as well.

One final note: The issue number referenced is for the whole epic which also calls for an "owner" mode, which wasn't implemented here as this requires a lot more changes and testing (as discussed on the mailing list), i.e. this ticket will not fully implement the JIRA ticket.

mposolda reviewed (approved):
@Martin-Idel-SI Thanks, I am approving this. The MigrationModelTest is not related to this PR, there is separate JIRA for that https://issues.redhat.com/browse/KEYCLOAK-13922 . So this PR should be fine from my PoV IMO.

stianst commented:
LGTM, great work @Martin-Idel-SI. Will give our writer a little bit of time to review the docs PR, once that's okay we will merge this stuff :)

BTW did you have plans to implement the owner sync-mode as well in a follow-up? I think that can be quite useful in the scenario where a user first registers with a social provider, but later decides to start managing their account through Keycloak instead.

Martin-Idel-SI commented:
@stianst I agree that it would be useful. I can't promise right now to follow up on this very soon since priorities are on a different topic now.
